### PR TITLE
React 18 support

### LINKED ISSUE #2800: Support React 18
# 🙋 Feature Request

We'd like to formally support React 18.

## 🤔 Expected Behavior

One thing we'll need to figure out is, for any of our containers (ie dialogs), do we need an API to mount the new/old way so we don't accidentally poison the downstream tree which may not be ready yet.

See this comment in particular https://github.com/adobe/react-spectrum/issues/2136#issuecomment-891374974

## 💁 Possible Solution

See Context.

## 🔦 Context

Starting and exploration work has been done here:
https://github.com/adobe/react-spectrum/pull/2526
https://github.com/adobe/react-spectrum/tree/react-18
https://github.com/adobe/react-spectrum/issues/2136

### PR DESCRIPTION
**Pull request by devongovett**
Closes #2526. Fixes #2800.

This adds basic support for React 18. Strict mode support will be done separately.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -150,17 +151,17 @@
...
+    "react-dom": "^18.0.0",
     "react-overlays": "0.8.3",
     "react-test-renderer": "^16.9.0",
     "recursive-readdir": "^2.2.2",
     "regenerator-runtime": "0.13.3",
     "rimraf": "^2.6.3",
     "sharp": "^0.29.3",
     "sinon": "^7.3.1",
-    "storybook-dark-mode": "^1.0.3",
+    "storybook-dark-mode": "^1.1.1-canary.120.3843.0",
```
[the remark this answers is not in the document]
Comment (devongovett) [reply package.json:164]: When I upgraded storybook to pre-release the command also updated this one

File: packages/@react-stately/data/test/useAsyncList.test.js [hunk 2]
```diff
@@ -199,14 +189,13 @@ describe('useAsyncList', () => {
...
   it('should return error in case fetch throws an error', async () => {
     let loadSpyThatThrows = jest.fn().mockRejectedValue(new Error('error'));
-    let {result, waitForNextUpdate} = renderHook(() => useAsyncList({
+    let {result} = renderHook(() => useAsyncList({
       load: loadSpyThatThrows
     }));
 
     expect(result.current.loadingState).toBe('loading');
 
     await act(async () => {
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-stately/data/test/useAsyncList.test.js:198]: hmm probably not. I don't remember adding it.

File: packages/@react-aria/menu/src/useMenuTrigger.ts [hunk 3]
```diff
@@ -91,6 +91,7 @@ export function useMenuTrigger(props: MenuTriggerAriaProps, state: MenuTriggerSt
 
   let formatMessage = useMessageFormatter(intlMessages);
   let {longPressProps} = useLongPress({
+    isDisabled: isDisabled || trigger !== 'longPress',
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-aria/menu/src/useMenuTrigger.ts:94]: Yeah I think I was debugging a test and noticed the unreleased long press event handlers being added

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (snowystinger) [reply package.json:164]: ah, didn't realize it was an automated upgrade, thanks for the explanation 👍🏻

File: packages/@adobe/react-spectrum/package.json [hunk 4]
```diff
@@ -67,7 +67,7 @@
     "access": "public"
   },
   "peerDependencies": {
-    "react": "^16.8.0 || ^17.0.0-rc.1",
-    "react-dom": "^16.8.0 || ^17.0.0-rc.1"
+    "react": "^16.8.0 || ^17.0.0-rc.1 || ^18.0.0",
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@adobe/react-spectrum/package.json:70]: Found what i was thinking of https://github.com/adobe/react-spectrum/issues/2054#issuecomment-868188590
it's closed, so we don't need to worry about this

**Review by snowystinger — DISMISSED** [review]
Code looks good, approving
Wait until testing session is done for merging

**Review by LFDanLu — APPROVED** [review]
Did a quick sweep of some components/functionality (liveannouncer, drag preview, etc) and things seems to work. Code looks fine to me as well. Looks like there are some conflicts now, but they seem innocuous (ListView one is just the `disabledBehavior` stuff from the disabled rows PR)

**Review by dannify — APPROVED** [review]

**Review by dannify — APPROVED** [review]

**Review by LFDanLu — APPROVED** [review]
reapproving, looks fine locally

**Pull request merged**